File: eulercycle.sol.java
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Queue;
import java.util.Stack;
import java.util.Comparator;
import java.lang.Math;


class Eulercycle {
    public Eulercycle() {};

    public int[] eulercycle(int num_node, int[][] edges) {
        return new int[]{0,1,2,0};
    }

    public boolean check(int[] path, int[][] edges) {
        if(path.length != edges.length + 1)
            return false;

        int[][] output_edges = new int[path.length - 1][2];
        for(int i=0; i<path.length-1; ++i) {
            output_edges[i][0] = Math.min(path[i], path[i+1]);
            output_edges[i][1] = Math.max(path[i], path[i+1]);
        }
        Arrays.sort(output_edges, (a, b) -> Arrays.compare(a,b));
        Arrays.sort(edges, (a, b) -> Arrays.compare(a,b));

        for(int i=0;i<path.length-1; ++i)
            if(!Arrays.equals(output_edges[i], edges[i]))
                return false;
        return true;
    }

    public static void main(String[] args) {
        Eulercycle m = new Eulercycle();
        int[][] edges = new int[][]{ {0,1}, {1,2}, {0,2} };
        int[] path = m.eulercycle(3, edges);
        // 0 1 2 0
        System.out.println(m.check(path, edges));
    }
}

File: eulercycle.sol.py
import sys 
from typing import List
sys.setrecursionlimit(100001)


class Eulercycle:
    def eulercycle(self, num_node:int, edges:List[int]) -> List[int]:
        # build edge list for nodes
        node_edges = [[] for i in range(num_node)]
        for i in range(len(edges)):
            node_edges[edges[i][0]].append(i)
            node_edges[edges[i][1]].append(i)

        # dfs
        used = [False] * len(edges)
        path = []
        def dfs(f):
            for i in node_edges[f]:
                if used[i]:
                    continue
                used[i] = True
                if edges[i][0] == f:
                    to = edges[i][1]
                else:
                    to = edges[i][0]
                dfs(to)
            path.append(f)

        dfs(0)
        print(path)
        return path


def toEdge(a):
    return [[min(a[i-1:i+1]), max(a[i-1:i+1])] for i in range(1, len(a))]


if __name__ == "__main__":
    edges = [[0,1],[1,2],[0,2]]
    path = Eulercycle().eulercycle(3, edges)
    # [0, 2, 1, 4, 3, 1, 0]
    print(len(path) == len(edges) + 1 and
          sorted(toEdge(path)) == sorted([[min(i), max(i)] for i in edges]))

    edges = [[0,1],[1,2],[0,2], [1,3],[1,4], [3,4]]
    path = Eulercycle().eulercycle(5, edges)
    # [0, 2, 1, 4, 3, 1, 0]
    print(len(path) == len(edges) + 1 and
          sorted(toEdge(path)) == sorted([[min(i), max(i)] for i in edges]))

    edges = [[0,1],[1,2],[0,2], [1,4],[1,3], [3,4]]
    path = Eulercycle().eulercycle(5, edges)
    print(len(path) == len(edges) + 1 and
          sorted(toEdge(path)) == sorted([[min(i), max(i)] for i in edges]))

    edges = [[0,1],[1,2],[0,2], [0,4],[0,3], [3,4]]
    path = Eulercycle().eulercycle(5, edges)
    # [0, 3, 4, 0, 2, 1, 0]
    print(len(path) == len(edges) + 1 and
          sorted(toEdge(path)) == sorted([[min(i), max(i)] for i in edges]))

    edges = [[0,1],[1,2],[0,2], [0,3],[0,4], [3,4]]
    path = Eulercycle().eulercycle(5, edges)
    print(len(path) == len(edges) + 1 and
          sorted(toEdge(path)) == sorted([[min(i), max(i)] for i in edges]))

    edges = [[0,1],[1,2],[0,2], [0,0]]
    path = Eulercycle().eulercycle(3, edges)
    # [0, 0, 2, 1, 0]
    print(len(path) == len(edges) + 1 and
          sorted(toEdge(path)) == sorted([[min(i), max(i)] for i in edges]))
